### PR TITLE
Improved IPv6 Support + Private Network Filtering

### PR DESCRIPTION
**Pull request by SyntaxNode**
Implements https://github.com/prebid/prebid-server/issues/1106 and the non-privacy features of https://github.com/prebid/prebid-server/issues/1332.

### CODE REVIEW

File: util/iputil/parse.go [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+const (
+	IPvUnknown IPVersion = 0
+	IPv4       IPVersion = 4
+	IPv6       IPVersion = 6
+)
+
+// ParseIP parses v as an ip address returning the result and version, or nil and unknown if invalid.
+func ParseIP(v string) (net.IP, IPVersion) {
+	if ip := net.ParseIP(v); ip != nil {
+		if strings.ContainsRune(v, ':') {
```
Comment (SyntaxNode) [inline util/iputil/parse.go:21]: net.ParseIP automatically converts IPv4 addresses to IPv6. There's no way to tell if the original value was IPv4 or IPv6 without the above check. This approach is common in other Go libraries.

File: config/config.go [hunk 2]
```diff
@@ -497,6 +499,21 @@ func New(v *viper.Viper) (*Configuration, error) {
...
+	if err := c.RequestValidation.Parse(); err != nil {
+		return nil, err
+	}
+
+	if err := isValidCookieSize(c.HostCookie.MaxCookieSizeBytes); err != nil {
+		glog.Fatal(fmt.Printf("Max cookie size %d cannot be less than %d \n", c.HostCookie.MaxCookieSizeBytes, MIN_COOKIE_SIZE_BYTES))
+		return nil, err
+	}
+
+	glog.Info("Logging the resolved configuration:")
```
Comment (hhhjort) [inline config/config.go:511]: Why did you move this logging above the map generation? This way the generated maps will not be included in the logs.

Comment (SyntaxNode) [reply config/config.go:511]: The logging call looks to correlate to the validate call. Such that, if validation fails you can refer to the logged configuration to figure out what went wrong. The hash sets don't affect validation, they're just an optimization. so I moved them below the validation call.

File: util/iputil/matcher.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package iputil
+
+import (
+	"net"
+)
+
+// IPMatcher is the interface for matching on an ip address and version.
+type IPMatcher interface {
```
Comment (hhhjort) [inline util/iputil/matcher.go:8]: I wonder if we should change "Match" to "Validate". We have the counterintuitive result that if `Match()` returns true, the IP was not on the list of IPs in the IPMatcher

Comment (SyntaxNode) [reply util/iputil/matcher.go:8]: Yeah. I cringed at that too. We're looking for public networks from a list of private networks, so there's got to be a negation somewhere. Are you suggesting only a name change or to flip the logic too?

I choose the matcher name by following the naming pattern used by the standard sort package. I'm sure having FindIP take an IPValidator works too. :)

Comment (hhhjort) [reply util/iputil/matcher.go:8]: I was just thinking of changing the name. If we flip the logic, then there is no need to change the name in my view.

Comment (SyntaxNode) [reply util/iputil/matcher.go:8]: Sure.

File: endpoints/openrtb2/auction.go [hunk 4]
```diff
@@ -914,13 +920,27 @@ func validateRegs(regs *openrtb.Regs) error {
...
 // setFieldsImplicitly uses _implicit_ information from the httpReq to set values on bidReq.
 // This function does not consume the request body, which was set explicitly, but infers certain
 // OpenRTB properties from the headers and other implicit info.
 //
 // This function _should not_ override any fields which were defined explicitly by the caller in the request.
 func (deps *endpointDeps) setFieldsImplicitly(httpReq *http.Request, bidReq *openrtb.BidRequest) {
-	setDeviceImplicitly(httpReq, bidReq)
+	sanitizeRequest(bidReq, deps.privateNetworkIPMatcher)
+
+	setDeviceImplicitly(httpReq, bidReq, deps.privateNetworkIPMatcher)
```
[the remark this answers is not in the document]
Comment (SyntaxNode) [reply endpoints/openrtb2/auction.go:943]: You're right, the information is available in the cfg object already. My thought in passing it down in the `endpointDep` is to avoid recreating the matcher object for each request, to save a bit on memory usage.

File: config/config.go (hunk 2 above)
Comment (SyntaxNode) [reply config/config.go]: I switched it back.

File: util/iputil/matcher.go (hunk 3 above)
Comment (SyntaxNode) [reply util/iputil/matcher.go]: Renamed to IPValidator and IPValidator.IsValid.
